wsgidav/addons/seafile/domain_controller.py

```diff
@@ -43,16 +56,33 @@ def authDomainUser(self, realmname, username, password, environ):
                 if not self.session_cls:
                     return False
                 # Assume that @username is a contact_email, get real email from seahub_db
+                email = None
                 session = self.session_cls()
                 profile_profile = Base.classes.profile_profile
                 q = session.query(profile_profile.user).filter(profile_profile.contact_email==username)
-                email = q.first()[0]
-                session.close()
+                res = q.first()
+                if res:
+                    email = res[0]
+                    if self.ccnet_threaded_rpc.validate_emailuser(email, password) != 0:
+                        email = None
+
+                # Assume that user is logging in with shibboleth, validate dedicated password from seahub_db
                 if not email:
-                    return False
-                if self.ccnet_threaded_rpc.validate_emailuser(email, password) != 0:
-                    return False
+                    secret = seahub_settings.SECRET_KEY[:BLOCK_SIZE]
+                    cipher = AES.new(secret)
+                    encoded_str = 'aes$' + EncodeAES(cipher, password)
```

Comment:
Why is it encrypted instead of hashed?

Reply:
This is why all other services add app passwords. Those can only be read once and are given a name which can be seen in the web interface. For a new login another app password is added and an outdated one can be deleted by the user.

With the readable password access to a victims account can be reached by stealing its token. The victim wouldn't notice anything. With an app password another app password would need to be added which is a least suspicious. It also allows locking out specific devices without having to completely change the dedicated password and locking out all devices.